arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/NonEmptyList.kt

```diff
@@ -182,46 +188,42 @@ public class NonEmptyList<out A>(
     NonEmptyList(all + a)
 
   public inline fun <B> foldLeft(b: B, f: (B, A) -> B): B =
-    this.tail.fold(f(b, this.head), f)
-
-  public fun <B> coflatMap(f: (NonEmptyList<A>) -> B): NonEmptyList<B> {
-    val buf = mutableListOf<B>()
-    tailrec fun consume(list: List<A>): List<B> =
-      if (list.isEmpty()) {
-        buf
-      } else {
-        val tail = list.subList(1, list.size)
-        buf += f(NonEmptyList(list[0], tail))
-        consume(tail)
+    all.fold(b, f)
+
+  public fun <B> coflatMap(f: (NonEmptyList<A>) -> B): NonEmptyList<B> =
+    buildList {
+      var current = all
+      while (current.isNotEmpty()) {
+        add(f(NonEmptyList(current)))
+        current = current.drop(1)
       }
-    return NonEmptyList(f(this), consume(this.tail))
-  }
+    }.let(::NonEmptyList)
 
   public fun extract(): A =
     this.head
 
-  override fun equals(other: Any?): Boolean =
-    super.equals(other)
-
-  override fun hashCode(): Int =
-    super.hashCode()
-
   override fun toString(): String =
     "NonEmptyList(${all.joinToString()})"
 
   public fun <B> align(b: NonEmptyList<B>): NonEmptyList<Ior<A, B>> =
-    NonEmptyList(Ior.Both(head, b.head), tail.align(b.tail))
+    NonEmptyList(all.align(b))
 
   public fun salign(SA: Semigroup<@UnsafeVariance A>, b: NonEmptyList<@UnsafeVariance A>): NonEmptyList<A> =
-    SA.run {
-      NonEmptyList(head.combine(b.head), tail.salign(SA, b.tail).toList())
-    }
+    NonEmptyList(all.salign(SA, b).toList())
 
   public fun <B> padZip(other: NonEmptyList<B>): NonEmptyList<Pair<A?, B?>> =
-    NonEmptyList(head to other.head, tail.padZip(other.tail))
+    NonEmptyList(all.padZip(other))
 
   public companion object {
 
+    @JvmStatic @JvmName("of")
+    public fun <A> of(first: A, vararg tail: A): NonEmptyList<A> =
+      NonEmptyList(listOf(first) + tail)
+
+    @JvmStatic @JvmName("ofOrNull")
+    public fun <A> ofOrNull(list: Iterable<A>): NonEmptyList<A>? =
+      list.firstOrNull()?.let { NonEmptyList(list.toList()) }
```

Comment:
Weren't these constructors deprecated in 1.x.x in favour of `nonEmptyListOf` to be in line with `listOf` etc from Kotlin Std?

Reply:
These were added back so that they could be called from Java, because mangling make it otherwise impossible. We could also remove them.

Reply:
🤔 I would be in favour to remove them. Optimising for Java seems strange, especially since on the Java side it just returns `List` anyway.